### PR TITLE
[AArch64][GlobalISel] Lower G_BUILD_VECTOR to G_INSERT_VECTOR_ELT

### PR DESCRIPTION
**Pull request by chuongg3**
The lowering happens in post-legalizer lowering if any source registers from G_BUILD_VECTOR are not constants.

Add pattern pragment setting `scalar_to_vector ($src)` asequivalent to `vector_insert (undef), ($src), (i61 0)`

Some of the regression are addressed in #105689 and #105692

### CODE REVIEW

File: llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp [hunk 1]
```diff
@@ -1048,6 +1048,45 @@ void applyLowerVectorFCMP(MachineInstr &MI, MachineRegisterInfo &MRI,
   MI.eraseFromParent();
 }
 
+// Intend to match the last part of
+// AArch64TargetLowering::LowerBUILD_VECTOR(SDValue Op, SelectionDAG &DAG)
+bool matchLowerBuildToInsertVecElt(MachineInstr &MI, MachineRegisterInfo &MRI) {
+  assert(MI.getOpcode() == TargetOpcode::G_BUILD_VECTOR &&
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:1054]: `GBuildVector *Build = cast<GBuildVector>(MI);`

File: llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp [hunk 2]
```diff
@@ -1048,6 +1048,45 @@ void applyLowerVectorFCMP(MachineInstr &MI, MachineRegisterInfo &MRI,
...
+         "Expected G_BUILD_VECTOR instruction");
+  bool isConstant = true;
+
+  // Check if the values are the same
+  for (unsigned i = 1; i < MI.getNumOperands(); i++) {
+    auto ConstVal =
+        getAnyConstantVRegValWithLookThrough(MI.getOperand(i).getReg(), MRI);
+    if (!ConstVal.has_value()) {
+      isConstant = false;
+    }
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:1064]: `break;`

Comment (tschuett) [reply llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:1064]: or `return false;`

File: llvm/test/CodeGen/AArch64/GlobalISel/legalize-shuffle-vector-widen-crash.ll [hunk 3]
```diff
@@ -10,12 +10,14 @@ define i32 @bar() {
...
-; CHECK-NEXT:    mov b2, v0[2]
-; CHECK-NEXT:    mov b3, v0[3]
-; CHECK-NEXT:    mov.h v0[1], v1[0]
-; CHECK-NEXT:    mov.h v2[1], v3[0]
+; CHECK-NEXT:    mov b2, v0[3]
+; CHECK-NEXT:    mov b3, v0[2]
+; CHECK-NEXT:    fmov w8, s1
+; CHECK-NEXT:    fmov w9, s2
+; CHECK-NEXT:    mov.h v0[1], w8
+; CHECK-NEXT:    mov.h v3[1], w9
```
Comment (aemerson) [inline llvm/test/CodeGen/AArch64/GlobalISel/legalize-shuffle-vector-widen-crash.ll:18]: ?!

File: llvm/lib/CodeGen/GlobalISel/GIMatchTableExecutor.cpp [hunk 4]
```diff
@@ -61,15 +61,41 @@ bool GIMatchTableExecutor::isBaseWithConstantOffset(
...
+      std::next(MI.getIterator()) == IntoMIIter)
     return true;
 
   // Convergent instructions cannot be moved in the CFG.
   if (MI.isConvergent() && MI.getParent() != IntoMI.getParent())
     return false;
 
-  return !MI.mayLoadOrStore() && !MI.mayRaiseFPException() &&
-         !MI.hasUnmodeledSideEffects() && MI.implicit_operands().empty();
+  if (MI.isLoadFoldBarrier())
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/GIMatchTableExecutor.cpp:75]: This PR has pulled in an entirely unrelated commit, any lowering patch shouldn't need to touch the selector tables

Comment (dc03-work) [reply llvm/lib/CodeGen/GlobalISel/GIMatchTableExecutor.cpp:75]: Looks like https://github.com/llvm/llvm-project/pull/105689 accidentally got included here.

Comment (chuongg3) [reply llvm/lib/CodeGen/GlobalISel/GIMatchTableExecutor.cpp:75]: I chose to include the TableGen patch to reduce the amount of regressions this patch creates to help with code review. This lowering patch was intended to help with selecting LD1 instructions but some of the patterns for LD1 were not matching due to reasons decribed in #101675.

File: llvm/test/CodeGen/AArch64/GlobalISel/legalize-shuffle-vector-widen-crash.ll (hunk 3 above)
Comment (davemgreen) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalize-shuffle-vector-widen-crash.ll:18]: When we went through all the changes in this pr, we decided this one should be using a vector any_ext. (And the zero's should all optimize away).

File: llvm/lib/Target/AArch64/AArch64Combine.td [hunk 5]
```diff
@@ -290,6 +290,13 @@ def combine_mul_cmlt : GICombineRule<
   (apply [{ applyCombineMulCMLT(*${root}, MRI, B, ${matchinfo}); }])
 >;
 
+def lower_build_insert_vec_elt : GICombineRule<
+  (defs root:$root, register_matchinfo:$matchinfo),
+  (match (G_BUILD_VECTOR $dst, GIVariadic<>:$unused):$root,
+        [{ return matchLowerBuildToInsertVecElt(*${root}, MRI); }]),
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64Combine.td:296]: I would add a couple of extra spaces to make the [{ inside the (

File: llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp [hunk 6]
```diff
@@ -1054,6 +1054,40 @@ void applyLowerVectorFCMP(MachineInstr &MI, MachineRegisterInfo &MRI,
   MI.eraseFromParent();
 }
 
+// Matches G_BUILD_VECTOR where at least one source operand is not a constant
+bool matchLowerBuildToInsertVecElt(MachineInstr &MI, MachineRegisterInfo &MRI) {
+  auto *GBuildVec = dyn_cast<GBuildVector>(&MI);
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:1059]: dyn_cast -> cast

File: llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp [hunk 7]
```diff
@@ -1054,6 +1054,40 @@ void applyLowerVectorFCMP(MachineInstr &MI, MachineRegisterInfo &MRI,
...
+
+  // Check if the values are all constants
+  for (unsigned i = 0; i < GBuildVec->getNumSources(); i++) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:1062]: i -> I

Comment (tschuett) [reply llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:1062]: `ì++` -> `++I`

File: llvm/lib/Target/AArch64/AArch64Combine.td [hunk 8]
```diff
@@ -290,6 +290,13 @@ def combine_mul_cmlt : GICombineRule<
   (apply [{ applyCombineMulCMLT(*${root}, MRI, B, ${matchinfo}); }])
 >;
 
+def lower_build_insert_vec_elt : GICombineRule<
```
[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/lib/Target/AArch64/AArch64Combine.td:293]: Bump on this, why is this an optimization combine and not in LegalizerHelper? The title says its lowering, but this is not

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/lib/Target/AArch64/AArch64Combine.td:293]: This doesn't seem good. Unrolling a build_vector into a sequence of vector inserts naturally belongs in LegalizerHelper

[the remark this answers is not in the document]
Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64Combine.td:293]: I'm not sure you can get through legalization without buildvector (or merge), without it making quite a mess. This fits in with the existing way that AArch64 handles shuffle vectors / buildvectors / other vector operations. It has lower in the name but it can be thought of as a combine that splits the buildvectors up prior to selection to help use the existing tablegen patterns.

**Review by davemgreen — APPROVED** [review]
This LGTM, and it seems to fit into the existing design of doing certain lowering post-legalization. We went through the test changes and decided they were improvements or should be optimized in some other way (such as #105692).

**Comment by kazutakahirata** [comment]
I've fixed a warning from this PR with a628bc3c2e7314e4b7c9af0d10cf39a70c731d15.

**Comment by chuongg3** [comment]
> I've fixed a warning from this PR with [a628bc3](https://github.com/llvm/llvm-project/commit/a628bc3c2e7314e4b7c9af0d10cf39a70c731d15).

Thank you!
